### DISCUSSION TITLE
Can't figure out how to create records for explicit many-to-many

### DISCUSSION THREAD

**Discussion by meotimdihia** [prisma/prisma#4089] [Q&A] [answered]
I'm new to Prisma. I Read:
https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/relation-queries#fluent-api
https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-client/crud#updatemany
https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#many-to-many-relations

Many to Many is the most complex relation, but the documentation doesn't have any example about how to create many-to-many in nodejs. 

I searched in issues and discussions too.
Can someone give me 1 code example about create explicit many-to-many ?

**Comment by ryands17** [comment, marked as answer]
Here's the docs on many-to-many relations: https://www.prisma.io/docs/reference/tools-and-interfaces/prisma-schema/relations#many-to-many-relations

I will add the same example here as in the docs for a better idea:

```prisma
model Post {
  id         Int                 @id @default(autoincrement())
  title      String
  categories CategoriesOnPosts[]
}
model Category {
  id    Int                 @id @default(autoincrement())
  name  String
  posts CategoriesOnPosts[]
}
model CategoriesOnPosts {
  post        Post     @relation(fields: [postId], references: [id])
  postId      Int       // relation scalar field (used in the `@relation` attribute above)
  category    Category @relation(fields: [categoryId], references: [id])
  categoryId  Int      // relation scalar field (used in the `@relation` attribute above)
  createdAt   DateTime @default(now())
  @@id([postId, categoryId])
}
```

This is how you can create an explicit m-n relation.

**Reply by meotimdihia** [reply]
@ryands17: I am trying to test in my existing databases. I knew how to create schemas. But I stuck at how to create data:  
```node
  async __savePeople(id, names: string[], type) {
    for (const name of names) {
      const result = await prisma.md_people.upsert({
        where: { name: name },
        create: { name: name, slug: slugify(name), type: type },
        update: { type: type }
      })
      // await prisma.md_people.update({
      //   where: {id: result.id},
      //   data: {
      //     md_comic_md_peoples: {
      //       connectOrCreate: {
      //         where: { A: id, type: type },
      //         create: {
      //           A: id, type: type
      //         },
      //       },
      //     }
      //   }
      // })
      

      await prisma.md_comic_md_peoples.create({ 
        data: { type: type},
        
      })
      debugger
    }
  }
```
![Code_2020-10-29_13-11-59](https://user-images.githubusercontent.com/300961/97532192-5be87480-19e8-11eb-9e4f-518c9ad3e65f.png)

![Code_2020-10-29_13-15-20](https://user-images.githubusercontent.com/300961/97532401-d4e7cc00-19e8-11eb-9202-ef8cbc4a403f.png)

**Reply by ryands17** [reply]
Taking this schema:

```prisma
model Post {
  id         Int                 @id @default(autoincrement())
  title      String
  categories CategoriesOnPosts[]
}
model Category {
  id    Int                 @id @default(autoincrement())
  name  String
  posts CategoriesOnPosts[]
}
model CategoriesOnPosts {
  post        Post     @relation(fields: [postId], references: [id])
  postId      Int       // relation scalar field (used in the `@relation` attribute above)
  category    Category @relation(fields: [categoryId], references: [id])
  categoryId  Int      // relation scalar field (used in the `@relation` attribute above)
  createdAt   DateTime @default(now())
  @@id([postId, categoryId])
}
```

I can create the relation as follows:

```ts
  const c1 = await prisma.category.create({ data: { name: 'category-1' } })
  const c2 = await prisma.category.create({ data: { name: 'category-2' } })

  const post = await prisma.post.create({ data: { title: 'post-1' } })

  // add categories to the post
  await prisma.post.update({
    where: { id: post.id },
    data: {
      categories: {
        create: [
          { category: { connect: { id: c1.id } } },
          { category: { connect: { id: c2.id } } },
        ],
      },
    },
  })
```

**Reply by meotimdihia** [reply]
Thanks for the guide, I can create it now.  

``` node
      await prisma.md_comics.update({
        where: { id: id },
        data: {
          md_people: {
            create: [
              {
                type: type,
                md_people: { connect: { id: result.id } }
              }
            ]
          }
        }
      })
```

![dbeaver_2020-10-29_13-58-27](https://user-images.githubusercontent.com/300961/97536144-4f1b4f00-19ef-11eb-90c3-37ac1daee6a6.png)

It's hard to figure it out if you don't show how to do it.

**Reply by ryands17** [reply]
Thanks for the feedback. We will add an example in the docs as well :)

**Reply by hardfire** [reply]
> Taking this schema:
> 
> ```prisma
> model Post {
>   id         Int                 @id @default(autoincrement())
>   title      String
>   categories CategoriesOnPosts[]
> }
> model Category {
>   id    Int                 @id @default(autoincrement())
>   name  String
>   posts CategoriesOnPosts[]
> }
> model CategoriesOnPosts {
>   post        Post     @relation(fields: [postId], references: [id])
>   postId      Int       // relation scalar field (used in the `@relation` attribute above)
>   category    Category @relation(fields: [categoryId], references: [id])
>   categoryId  Int      // relation scalar field (used in the `@relation` attribute above)
>   createdAt   DateTime @default(now())
>   @@id([postId, categoryId])
> }
> ```
> 
> I can create the relation as follows:
> 
> ```ts
>   const c1 = await prisma.category.create({ data: { name: 'category-1' } })
>   const c2 = await prisma.category.create({ data: { name: 'category-2' } })
> 
>   const post = await prisma.post.create({ data: { title: 'post-1' } })
> 
>   // add categories to the post
>   await prisma.post.update({
>     where: { id: post.id },
>     data: {
>       categories: {
>         create: [
>           { category: { connect: { id: c1.id } } },
>           { category: { connect: { id: c2.id } } },
>         ],
>       },
>     },
>   })
> ```

I'm running almost similar query and get the following error when running the query above (with only one category to link to)

```
Argument categoryId for data.categories.create.0.categoryId is missing.                                                 
Argument postId for data.categories.create.0.postId is missing.    
```

And if i add these to the object then i get the following error

```
Invalid `prisma.post.update()` invocation:                                                                                                                                                                                               
                                                                                                                                                                                                                                               
                                                                                                                                                                                                                                               
  PANIC: called `Option::unwrap()` on a `None` value                                                                                                                                                                                           
                                                                                                                                                                                                                                               
This is a non-recoverable error which probably happens when the Prisma Query Engine has a panic. 
```

**Reply by meotimdihia** [reply]
@hardfire tips to write code with Prisma: install autocomplete if you use VScode: https://marketplace.visualstudio.com/items?itemname=prisma.prisma

**Reply by ryands17** [reply]
@hardfire Could you share your schema and query? As per the error it seems that the `categoryId` and/or the `postId` is undefined while being passed.

**Reply by hardfire** [reply]
Ok, so my schema/query has Collections and Todos (instead of posts and categories). This is what the schema looks like

```prisma
model Collection {
  id                Int           @id @default(autoincrement())
  createdAt         DateTime      @default(now())
  updatedAt         DateTime      @updatedAt
  todos             TodosInCollection[]
}

model Todo {
  id                Int         @id @default(autoincrement())
  title             String
  collections       TodosInCollection[]
}

model TodosInCollection {
  collection      Collection @relation(fields: [collectionId], references: [id])
  collectionId    Int
  todo            Todo @relation(fields: [todoId], references: [id])
  todoId          Int
  createdAt       DateTime @default(now())
  @@id([todoId, collectionId])
}

```

and my query is the following
```ts
  const updatedCollection = await db.collection.update({
    where: { id: collectionId},
    data: {
      todos: {
        create: [{ 
            todo: { 
              connect: { 
                id: todoId
              },
            },
        }],
      },
    },
  })

```

**Reply by ryands17** [reply]
@hardfire The below query works fine on Prisma version _2.12.1_.

```ts
const { id: collectionId } = await prisma.collection.create({ data: {} })
  const { id: todoId } = await prisma.todo.create({ data: { title: 'Todo 1' } })

  const updatedCollection = await prisma.collection.update({
    where: { id: collectionId },
    data: {
      todos: {
        create: [
          {
            todo: {
              connect: {
                id: todoId,
              },
            },
          },
        ],
      },
    },
  })
```

Could you check if the `todoId` and `collectionId` are defined and exist in the DB?

**Reply by hardfire** [reply]
I just checked and the two fields are defined in the DB. 
![image](https://user-images.githubusercontent.com/513457/100742955-b586f900-3403-11eb-8474-2de15ed4e53b.png)

I also upgraded to prisma 2.12.1 to be sure and still get the same error.

**Reply by hardfire** [reply]
Ok, Generating the prisma client with 2.12.1 now fixed the issue. 
Sorry for the hassle and thank you for helping out :D.
I was in 2.12.0 earlier

**Reply by hardfire** [reply]
Since most of the context is already here, i have a follow up question
running the following query runs the following queries in postgres
 - select collections table for collectionId
 - select todos table for todoId
 - insert into the many-many table
 - select from the many-many table.

I find this a little unoptimized for my usecase, so i'm curious if we can disable all the selects and run only a single insert query? 

```prisma
  const updatedCollection = await prisma.collection.update({
    where: { id: collectionId },
    data: {
      todos: {
        create: [
          {
            todo: {
              connect: {
                id: todoId,
              },
            },
          },
        ],
      },
    },
  })
```

**Reply by ryands17** [reply]
Hey @hardfire 👋 

Thanks for bringing this up. We will look into this as there are a couple of issues related to performance and reduction of queries, and we will surely be optimizing those in the future. The best workaround currently will be to use raw queries for the time being :)

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
